tests/components/enphase_envoy/test_switch.py

```diff
@@ -0,0 +1,206 @@
+                Platform.SWITCH,
+                option,
+                {ATTR_ENTITY_ID: test_entity},
+                blocking=True,
+            )
+        assert mock_go_on_grid.await_count == (2 if grid_status == STATE_OFF else 1)
+        assert mock_go_off_grid.await_count == (2 if grid_status == STATE_ON else 1)
+
+
+@pytest.mark.parametrize(("mock_envoy"), *ALL_FIXTURES, indirect=["mock_envoy"])
+async def test_switch_grid_charge(
+    hass: HomeAssistant,
+    config_entry: MockConfigEntry,
+    mock_envoy: AsyncMock,
+    mock_disable_charge_from_grid: AsyncMock,
+    mock_enable_charge_from_grid: AsyncMock,
+) -> None:
+    """Test enphase_envoy switch charge from grid operation."""
+
+    await setup_with_selected_platforms(hass, config_entry, [Platform.SWITCH])
+
+    # build switching orders
+    INITIAL_OFF_ORDER = (SERVICE_TURN_ON, SERVICE_TURN_OFF, SERVICE_TOGGLE)
+    INITIAL_ON_ORDER = (SERVICE_TURN_OFF, SERVICE_TURN_ON, SERVICE_TOGGLE)
+
+    entity_base = f"{Platform.SWITCH}.enpower_"
+
+    if (
+        mock_envoy.data.tariff
+        and mock_envoy.data.tariff.storage_settings
+        and mock_envoy.data.enpower
+    ):
+        sn = mock_envoy.data.enpower.serial_number
+        test_entity = f"{entity_base}{sn}_charge_from_grid"
+        charge_status = (
+            STATE_ON
+            if (mock_envoy.data.tariff.storage_settings.charge_from_grid)
+            else STATE_OFF
+        )
+        assert charge_status == hass.states.get(test_entity).state
+
+        # test charge from grid switch change operation
+        for option in (
+            INITIAL_OFF_ORDER if charge_status == STATE_OFF else INITIAL_ON_ORDER
+        ):
+            await hass.services.async_call(
+                Platform.SWITCH,
+                option,
+                {ATTR_ENTITY_ID: test_entity},
+                blocking=True,
+            )
+
+        assert mock_enable_charge_from_grid.await_count == (
+            2 if charge_status == STATE_OFF else 1
+        )
+        assert mock_disable_charge_from_grid.await_count == (
+            2 if charge_status == STATE_ON else 1
+        )
+
+        mock_enable_charge_from_grid.reset_mock()
+        mock_disable_charge_from_grid.reset_mock()
+
+
+@pytest.mark.parametrize(("mock_envoy"), *ALL_FIXTURES, indirect=["mock_envoy"])
+async def test_switch_relay_operation(
+    hass: HomeAssistant,
+    config_entry: MockConfigEntry,
+    mock_envoy: AsyncMock,
+    mock_close_dry_contact: AsyncMock,
+    mock_open_dry_contact: AsyncMock,
+) -> None:
+    """Test enphase_envoy relay entities operation."""
+
+    await setup_with_selected_platforms(hass, config_entry, [Platform.SWITCH])
+
+    # build switching orders
+    INITIAL_OFF_ORDER = (SERVICE_TURN_ON, SERVICE_TURN_OFF, SERVICE_TOGGLE)
+    INITIAL_ON_ORDER = (SERVICE_TURN_OFF, SERVICE_TURN_ON, SERVICE_TOGGLE)
+
+    entity_base = f"{Platform.SWITCH}."
+
+    if mock_envoy.data.dry_contact_status:
```

Comment:
Please avoid branching in tests.  In general we don't want to see `if` blocks in the body of a test

Reply:
Understand and I've tried it. Trying to strike a balance between creating and maintaining a large bunch of fixture files or a base set reflecting well known Envoy types. Choose for the latter, created topic tests but that required 1 if test to make sure it reflected the actual tests in pyenphase.